packages/ses/src/make-evaluate.js

```diff
@@ -31,80 +35,38 @@ function buildOptimizer(constants, name) {
-  //       trigger direct eval. The direct eval semantics is what allows the
-  //       evaluated code to lookup free variable names on the other scope
-  //       objects and not in global scope.
-  //    b) `moduleLexicals` which provide a way to introduce free variables
-  //       that are not available on the globalObject.
-  //    c) `globalObject` is the global scope object of the evaluator, aka the
-  //       Compartment's `globalThis`.
-  //    d) `scopeTerminator` is a proxy object which prevents free variable
-  //       lookups to escape to the start compartment's global object.
-  // 2: `optimizer`s catch constant variable names for speed.
-  // 3: The inner strict `evaluate` function should be passed two parameters:
-  //    a) its arguments[0] is the source to be directly evaluated.
-  //    b) its 'this' is the this binding seen by the code being
-  //       directly evaluated (the globalObject).
-
-  // Notes:
-  // - The `optimizer` strings only lookup values on the `globalObject` and
-  //   `moduleLexicals` objects by construct. Keywords like 'function' are
-  //   reserved and cannot be used as a variable, so they are excluded from the
-  //   optimizer. Furthermore to prevent shadowing 'eval', while a valid
-  //   identifier, that name is also explicitly excluded.
-  // - when 'eval' is looked up in the `evalScope`, the powerful unsafe eval
-  //   intrinsic is returned after automatically removing it from the
-  //   `evalScope`. Any further reference to 'eval' in the evaluate string will
-  //   get the tamed evaluator from the `globalObject`, if any.
+  return function evaluate(code) {
+    'use strict';
 
-  // TODO https://github.com/endojs/endo/issues/816
-  // The optimizer currently runs under sloppy mode, and although we doubt that
-  // there is any vulnerability introduced just by running the optimizer
-  // sloppy, we are much more confident in the semantics of strict mode.
-  // The `evaluate` function can be and is reused across multiple evaluations.
-  // Since the optimizer should not be re-evaluated every time, it cannot be
-  // inside the `evaluate` closure. While we could potentially nest an
-  // intermediate layer of `() => {'use strict'; ${optimizers}; ...`, it
-  // doesn't seem worth the overhead and complexity.
-  const evaluateFactory = FERAL_FUNCTION(`
-    with (this.scopeTerminator) {
-      with (this.globalObject) {
-        with (this.moduleLexicals) {
-          with (this.evalScope) {
-            ${globalObjectOptimizer}
-            ${moduleLexicalOptimizer}
-            return function() {
-              'use strict';
-              return eval(arguments[0]);
-            };
-          }
-        }
-      }
-    }
-  `);
+    context.evalScope.eval; // required by make-safe-evaluator.
 
-  return apply(evaluateFactory, context, []);
+    return apply(
+      FERAL_FUNCTION(
+        `{${SCOPE_TERMINATOR_KEYS}}`,
```

Comment:
I'm confused, how does it work for changing keys of the global object? Either the code changing a pre-existing global binding `foo = 42`, or someone changing an existing property of the global object and expecting the global binding to be updated. `globalThis.bar = 42; console.log(bar)`.

Reply:
There is currently no way to hook into scope binding access (without `this`). Someday we may have [`let` decorators](https://github.com/tc39/proposal-decorators/blob/master/EXTENSIONS.md#let-decorators) that would do that. The [refs proposal](https://github.com/rbuckton/proposal-refs) might also be able to do something here. Both of those would still require a non-extensible global object, but at least the bindings could be mutable (mutated either by the evaluated code, or reflect mutation of the source)

Reply:
We're talking about propagating updates between multiple evaluations within one compartment here I assume. So if we only ever had one, that would not be a problem. It'd be hard to pull off, but theoretically possible with a bundler.

Otherwise, manual propagation of individual globals seems to be the only option. 
A code snippet that reads specified local variables and passes them on to a callback on demand could work.